src/V3AstNodeExpr.h

```diff
@@ -69,6 +69,8 @@ class AstNodeBiop VL_NOT_FINAL : public AstNodeExpr {
     // Binary expression
     // @astgen op1 := lhsp : AstNodeExpr
     // @astgen op2 := rhsp : AstNodeExpr
+    bool m_pure;  // Cached results of isPure method
+    bool m_pureComputed = false;  // True if m_pure was already computed;
```

Comment:
Perhaps use a uint8_t with three states, to save a byte on all these common node types?

Should clone() clear m_pureComputed? For that matter mightn't a node edit somewhere down the tree affect this caching? Let's see if that's true or not, please have V3Broken recompute m_pure and assert if it is incorrectly cached (without going O(n^2)).

Reply:
I added V3Broken checks and a few tests failed. I will check why.

Reply:
I investigated the fail of `t_queue.pl`. It failed due to conversion in V3Premit.
I reproduced the fail on:
```systemverilog
module t (/*AUTOARG*/
   // Inputs
   clk
   );
   input clk;

   typedef integer q_t[$];

   always @ (posedge clk) begin
         typedef struct packed {
            bit [7:0] opcode;
            bit [127:0] data;
         } instructionW; // named structure type

         instructionW q[$];

         do if (q[0].data !== 0); while (0);
      $write("*-* All Finished *-*\n");
      $finish;
   end

endmodule
```
Both sides of `q[0].data !== 0` were replaced by temporary variables.
The AST before V3Premit:
```
    1:2:3: IF 0x555556d7cd80 <e4173> {d23an}
    1:2:3:1: NEQ 0x555556d546e0 <e5416#> {d23bb} @dt=0x555556e82820@(G/wu32/1)
    1:2:3:1:1: CONST 0x555556d56e00 <e2029> {d23bf} @dt=0x555556d91d40@(G/w128)  128'h0
    1:2:3:1:2: SEL 0x555556d7ccc0 <e2030> {d23aw} @dt=0x555556d91450@(G/w128)
    1:2:3:1:2:1: CMETHODHARD 0x555556e32700 <e5415#> {d23as} @dt=0x555556e82b60@(G/wu160/136)  at
    1:2:3:1:2:1:1: VARREF 0x555556e24690 <e1304> {d23ar} @dt=0x555556d7ca80@(w0)[queue]  t__DOT__unnamedblk1__DOT__q [RV] <- VAR 0x555556e00180 <e1892> {d21ax} @dt=0x555556d7ca80@(w0)[queue]  t__DOT__unnamedblk1__DOT__q [VSTATIC]  VAR
    1:2:3:1:2:1:2: CONST 0x555556d56f00 <e1560> {d23at} @dt=0x555556e20270@(G/sw32)  32'sh0
    1:2:3:1:2:2: CONST 0x555556d57000 <e1326> {d23aw} @dt=0x555556d91ba0@(G/w32)  32'h0
    1:2:3:1:2:3: CONST 0x555556d57100 <e1327> {d23aw} @dt=0x555556d91ba0@(G/w32)  32'h80
```
AST after V3Premit:
```
    1:2:2: VAR 0x555556d6c600 <e5536#> {d23bf} @dt=0x555556d91d40@(G/w128)  __Vtemp_1 STMTTEMP
    1:2:2: VAR 0x555556d6c780 <e5550#> {d23aw} @dt=0x555556d91450@(G/w128)  __Vtemp_2 STMTTEMP
    1:2:3: ASSIGN 0x555556d694a0 <e5542#> {d23bf} @dt=0x555556d91d40@(G/w128)
    1:2:3:1: CONST 0x555556d56e00 <e5540#> {d23bf} @dt=0x555556d91d40@(G/w128)  128'h0
    1:2:3:2: VARREF 0x555556d6b1d0 <e5541#> {d23bf} @dt=0x555556d91d40@(G/w128)  __Vtemp_1 [LV] => VAR 0x555556d6c600 <e5536#> {d23bf} @dt=0x555556d91d40@(G/w128)  __Vtemp_1 STMTTEMP
    1:2:3: ASSIGN 0x555556d69550 <e5556#> {d23aw} @dt=0x555556d91450@(G/w128)
    1:2:3:1: SEL 0x555556d7ccc0 <e5554#> {d23aw} @dt=0x555556d91450@(G/w128)
    1:2:3:1:1: CMETHODHARD 0x555556e32700 <e5415> {d23as} @dt=0x555556e82b60@(G/wu160/136)  at
    1:2:3:1:1:1: VARREF 0x555556e24690 <e1304> {d23ar} @dt=0x555556d7ca80@(w0)[queue]  t__DOT__unnamedblk1__DOT__q [RV] <- VAR 0x555556e00180 <e1892> {d21ax} @dt=0x555556d7ca80@(w0)[queue]  t__DOT__unnamedblk1__DOT__q [VSTATIC]  VAR
    1:2:3:1:1:2: CONST 0x555556d56f00 <e1560> {d23at} @dt=0x555556e20270@(G/sw32)  32'sh0
    1:2:3:1:2: CONST 0x555556d57000 <e1326> {d23aw} @dt=0x555556d91ba0@(G/w32)  32'h0
    1:2:3:1:3: CONST 0x555556d57100 <e1327> {d23aw} @dt=0x555556d91ba0@(G/w32)  32'h80
    1:2:3:2: VARREF 0x555556d6b3b0 <e5555#> {d23aw} @dt=0x555556d91450@(G/w128)  __Vtemp_2 [LV] => VAR 0x555556d6c780 <e5550#> {d23aw} @dt=0x555556d91450@(G/w128)  __Vtemp_2 STMTTEMP
    1:2:3: IF 0x555556d7cd80 <e5557#> {d23an}
    1:2:3:1: NEQ 0x555556d546e0 <e5416> {d23bb} @dt=0x555556e82820@(G/wu32/1)
    1:2:3:1:1: VARREF 0x555556d6b2c0 <e5546#> {d23bf} @dt=0x555556d91d40@(G/w128)  __Vtemp_1 [RV] <- VAR 0x555556d6c600 <e5536#> {d23bf} @dt=0x555556d91d40@(G/w128)  __Vtemp_1 STMTTEMP
    1:2:3:1:2: VARREF 0x555556d6b4a0 <e5560#> {d23aw} @dt=0x555556d91450@(G/w128)  __Vtemp_2 [RV] <- VAR 0x555556d6c780 <e5550#> {d23aw} @dt=0x555556d91450@(G/w128)  __Vtemp_2 STMTTEMP
```
Generally, the operation is pure. If `isPure` was run at an early phase of verilation, it would return `true`. Then, a select operation is replaced with `AstCMethodHard`. After the conversion, `isPure` returns `false`, because `AstCMethodHard` is treated as `impure`. So here the cached value would be wrong. In our example we didn't have that problem, because `isPure` wasn't run before the conversion and there was no cached value.

The purity was computed on some stage after the conversion and the cached value was `IMPURE`.
Then the conversion in V3Premit extracted the operands of `!==` to temporary variables. AstVarRefs are pure, so the purity has changed and the error was thrown.

In this example, the conversion in V3Premit isn't incorrect. But it would be incorrect if we had something like `X || (Y !== Z)` where Y or Z have side effects. In such a case, we can't move the operands of `!==` before the expression. So I think that this phase requires a change. Maybe it can be fixed by wrapping these operands into `AstExprStmt`, but maybe they just don't need to be converted. It has to be checked

So I found two problems:
* the possibility of change of purity from `pure` to `impure` when queue selects are replaced with `AstCMethodHard` (it can be fixed by overriding `isPure` in that class)
* the conversion in V3Premit that may break short-circuit evaluation.

Reply:
The other test that failed is `t_flag_comp_limit_parens`. It also failed due to a replacement of AstCMethodHard with a var ref.

Reply:
I think that non-void AstCMethodHard nodes are always pure. Void methods should rather be considered as statements and we shouldn't check their purity (since they don't return values, they probably always have side effects).

Reply:
CMethodHard is used also to e.g. modify associative containers and certainly can be impure. We probably need to mark each one when it's created as pure or not.  (Perhaps better as a separate pull?)

Reply:
Are there any methods that modify containers and are non-void?
First I wanted to define a set of impure methods, but any method I found was pure (or void, such as `sort` and such methods cannot be used in expression).

Reply:
Queue pop_front and pop_back.  If you want I'll audit more precisely and add isPure tags as needed (likely will default to impure and add tags everywhere else as is safer).

Reply:
You're right. There were some other methods too. I fixed the implementation.